app/src/org/commcare/android/framework/SessionActivityRegistration.java

```diff
@@ -65,10 +66,15 @@ public static void registerSessionExpiration() {
         }
     }
 
-    public static void returnToLogin(Context context) {
-        Intent i = new Intent(context.getApplicationContext(), LoginActivity.class);
-        i.setFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
-        i.putExtra(LoginActivity.REDIRECT_TO_HOMESCREEN, true);
+    private static void letHomeScreenRedirectToLogin(Context context) {
+        Intent i = new Intent(context.getApplicationContext(), CommCareHomeActivity.class);
+        i.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
         context.startActivity(i);
     }
+
+    public static void returnToLogin(Activity activity) {
+        Intent i = new Intent(activity.getApplicationContext(), LoginActivity.class);
+        i.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+        activity.startActivityForResult(i, CommCareHomeActivity.LOGIN_USER);
```

Comment:
Allows the launching activity to inspect the result of the ```LoginActivity```.
In the case of ```CommCareHomeActivity```, it will call ```finish``` if the result is cancel, and otherwise sync and show demo user warnings
see https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/dalvik/activities/CommCareHomeActivity.java#L573-L595